### PR TITLE
IntLayout try (WIP) + LongLayout + EnumLayout

### LINKED ISSUE #1811: Layoutable non-string values
Current approach:

render, parse.

e.g.

`Convert.ToInt64(this.MaxKilobytes.Render(logEvent), CultureInfo.InvariantCulture);`

This isn't optimal. 

Proposal, replace Layout with `Layout<int>` ~`IntLayout`,~ and let `render` do a parse.

related #1234 

update: We have now also an implicit conversion from `string` to `layout`. For the `int` case it should be also from `int` to `layout` (fixed values), hence we need `IntLayout` or something like that.

### PR DESCRIPTION
**Pull request by 304NotModified**
fixes #1811 related #1234 

todo

- [x] impl when parsing -> isFixed
- [x] trygetRawValue
- [x] null check/handling
- [x] generic/abstract, for boolLayout, longLayout, enumLayout etc.
- [ ] more unit tests
- [ ] update XSD generation stuff (fails appveyor now)
- [ ] More layouts
- [ ]  test with collections
- [ ] test with enum

choices:

- [ ] rename to LayoutableInt? (but need also LayoutableText/LayoutableString) -> maybe later

### CODE REVIEW

**Comment by 304NotModified** [comment]
besides tests, this is ready for int for now (MVP)

File: src/NLog/Layouts/Typed/LongLayout.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+using System.Globalization;
+using System.Linq;
+using NLog.Common;
+
+namespace NLog.Layouts
+{
+    /// <summary>
+    /// Layout rendering to Long
+    /// </summary>
+    public class LongLayout : TypedLayout<long?>
```
Comment (304NotModified) [inline src/NLog/Layouts/Typed/LongLayout.cs]: @snakefoot maybe this is a confusing name :(

**Comment by snakefoot** [comment]
Maybe just make a generic LayoutT<primitive> for all none-strings? Again mostly for intellisense


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


@304NotModified commented on this pull request.

________________________________

In src/NLog/Layouts/Typed/LongLayout.cs<https://github.com/NLog/NLog/pull/3119#discussion_r283015213>:

> +// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+using System;
+using System.ComponentModel;
+using System.Globalization;
+using System.Linq;
+using NLog.Common;
+
+namespace NLog.Layouts
+{
+    /// <summary>
+    /// Layout rendering to Long
+    /// </summary>
+    public class LongLayout : TypedLayout<long?>


@snakefoot<https://github.com/snakefoot> maybe this is a confusing name :(

**Comment by 304NotModified** [comment]
> Maybe just make a generic LayoutT<primitive>

Was also thinking about this, but it's a bit hard as we can't use `int.TryParse()` and `Convert.ToInt32()` etc

**Comment by snakefoot** [comment]
Could have restriction of using interface IConvertible or something else


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


Maybe just make a generic LayoutT

Was also thinking about this, but it's a bit hard as we can't use int.TryParse() and Convert.ToInt32() etc

**Comment by 304NotModified** [comment]
maybe rename TypedLayout<T> to SimpleLayout<T>?

File: src/NLog/Layouts/Typed/PrimitiveLayout.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+using System.ComponentModel;
+using System.Globalization;
+
+namespace NLog.Layouts
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    [CLSCompliant(false)]
```
Comment (304NotModified) [inline src/NLog/Layouts/Typed/PrimitiveLayout.cs]: @snakefoot this is a bit pity. `IConvertible` isn't CLSCompliant

Comment (304NotModified) [reply src/NLog/Layouts/Typed/PrimitiveLayout.cs]: and now we have also:

```
public class EventLogTarget
{
        ....
   
        [CLSCompliant(false)]
        public PrimitiveLayout<int> EventId { get; set; }

```

File: src/NLog/Layouts/Typed/SimpleLayout.cs [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+        }
+
+        /// <inheritdoc />
+        protected override bool TryParse(string text, out T value)
+        {
+            return TryConvertTo(text, out value);
+        }
+
+        /// <inheritdoc />
+        protected override bool TryConvertTo(object raw, out T value)
```
Comment (304NotModified) [inline src/NLog/Layouts/Typed/SimpleLayout.cs]: @snakefoot maybe this? I could drop typedlayout now.

File: src/NLog/Layouts/Typed/Layoutable.cs [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+        protected override string TypedName => typeof(T).Name; //todo cache?
+
+        /// <inheritdoc />
+        protected override string ValueToString(T value, CultureInfo cultureInfo)
+        {
+            if (value is IConvertible convertible)
+            {
+                return convertible.ToString(cultureInfo);
+            }
+
```
Comment (304NotModified) [inline src/NLog/Layouts/Typed/Layoutable.cs]: todo needs enum handling here?

File: src/NLog/Layouts/Typed/Layoutable.cs [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+            try
+            {
+                // todo could be better?
+                var objTypeCode = (default(T) as IConvertible)?.GetTypeCode();
+                switch (objTypeCode)
+                {
+                    //todo ugly :P
+                    case TypeCode.Boolean:
```
Comment (304NotModified) [inline src/NLog/Layouts/Typed/Layoutable.cs]: todo needs enum handling here?

File: src/NLog/Layouts/Typed/Layoutable.cs [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+                        {
+                            if (bool.TryParse(raw.ToString(), out var value1))
+                            {
+                                value = (T)(object)value1;
```
Comment (304NotModified) [inline src/NLog/Layouts/Typed/Layoutable.cs]: Is this a box/unbox operation?

Comment (304NotModified) [reply src/NLog/Layouts/Typed/Layoutable.cs]: Related https://stackoverflow.com/questions/1189144/c-sharp-non-boxing-conversion-of-generic-enum-to-int#23391746

**Comment by snakefoot** [comment]
Maybe call it `TypedLayout<T>` and let it support `TypeConverter`.

**Comment by 304NotModified** [comment]
> support `TypeConverter`.

or PropertyTypeConverter?

> Maybe call it `TypedLayout<T>`

What about `Layout<T> `?

**Comment by snakefoot** [comment]
Would it not conflict with non generic Layout?


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


support TypeConverter.

or PropertyTypeConverter?

Maybe call it TypedLayout<T>

What about Layout<T> ?

**Comment by 304NotModified** [comment]
nope, e.g. Tuple class

image

**Comment by 304NotModified** [comment]
superseded by https://github.com/NLog/NLog/pull/3656

**Pull request closed without merging**